README.md
## FPM-PLUGIN-ADMIN
用于显示后台管理页面的插件

### Install
```bash
yarn add fpm-plugin-admin
```

### Useage

default user
admin
741235896

可以在 config.json 中添加 admin 节点
```javascript
{
    //...
    "admin":{
        "user":"admin",
        "pass":"123123",
        "error":"用户名或密码错误"
    }
    //...
}
```

CHANGELOG.md
## 1.0.6
- BugFix
  - `fpmc.init({ ... master*K*ey: ...})` The `masterKey` the `K` should be upcase
## 1.0.3
- Add
  - `ctx.state` is the common variable object, it can be readed in view page

## 1.0.2
- Add
  - Show Plugins Detail

## 0.0.1

TODO: 
- Add logger Monitor
  - tail the app.log 
  - read from the end of the app.log
  - support the download link of the app.log

- Add `Extend` Menus [ Hard Thing ]
  - added from other plugins


test/function.test.js
var should = require("chai").should();
var YF = require("yf-fpm-client-nodejs").default;

YF.init({appkey: '123123', masterKey: '123123', endpoint: 'http://localhost:9999/api'});


describe('Function', function(){
  beforeEach(done => {
    done()
  })
  

  afterEach(done => {
    done()
  })

  it('Function A', function(done){
    var func = new YF.Func('test.foo');
    func.invoke({})
      .then(function(data){
        console.log(data)
        done();
      }).catch(function(err){
        done(err);
      })
  })
})


src/session.js
import path from 'path'
import process from 'process'
import _ from 'lodash'

const LOCAL = path.join(__dirname, '..')
const CWD = process.cwd()
const DEV_MODE = LOCAL == CWD
let VIEW_ROOT_DIR = ''
if(!DEV_MODE){
    VIEW_ROOT_DIR = 'node_modules/fpm-plugin-admin/'
}

export default async (ctx, next) => {
  let req = ctx.request
  let url = req.url
  if(_.startsWith(url, '/admin')){
    _.extend(ctx.state, { 
      env: ctx.fpm._env,
      currentlink: url,
      version: ctx.fpm.getVersion(),
      view_root_dir: VIEW_ROOT_DIR,
    } )
    if (url === '/admin' || url === '/admin/login') {
      await next()
      return
    }
    let admin = ctx.session.admin
    
    //TODO RBAC
    if (!admin) {
      ctx.redirect('/admin/login')
    } else {
      await next()
    }
  }else{
    await next()
  }
}

public/js/app.js
var $ = window.$ = (function(ctx){
  function _miniJq(selector){
    if(selector !== undefined){
      switch(typeof selector){
        case 'string':
          this._nodes = document.querySelectorAll(selector); 
          break;
        case 'object':
          this._nodes = [ selector ]
          break;
      } 
        
      this.length = this._nodes.length;
      if(this.length === 1){
        this._dom = this._nodes[0]
      }
    }
  }
  var miniJq = function(selector){
    return new _miniJq(selector);
  }
  miniJq.post = function(url, data){
    return axios.post(url, data)
            .then(function(rsp){
              return rsp.data;
            });
  };
  miniJq.get = function(url, data){
    return axios.get(url, data);
  };
  _miniJq.prototype = {
    val: function(v){
      if(v === undefined){
        return this._dom.value;
      }
      this._dom.value = v;
      return this;
    },
    attr: function(k, v){
      if(v === undefined){
        return this._dom.getAttribute(k);
      }
      this._dom.setAttribute(k, v);
      return this;
    },
    on: function(event, callback){
      _.map(this._nodes, function(node){
        node.addEventListener(event, callback, false);
      });
    },
    hasClass: function(clz){
      return !!this._dom.className.match( new RegExp( "(\\s|^)" + clz + "(\\s|$)") );
    },
    addClass: function(clz){
      this._dom.className += ' ' + clz;
      return this;
    },
    removeClass: function(clz){
      if(this.hasClass(clz)){
        this._dom.className = this._dom.className.replace( new RegExp( "(\\s|^)" + clz + "(\\s|$)" ), " " );
      }
      return this;
    },
    html: function(html){
      this._dom.innerHTML = html;
      return this;
    },
    show: function(){
      _.map(this._nodes, function(node){
        node.style.display = 'block';
      });      
      return this;
    },
    hide: function(){
      _.map(this._nodes, function(node){
        node.style.display = 'none';
      });
      return this;
    }
  }
  ctx.$ = miniJq;
  return miniJq;
})(window)

src/admin.js
import _ from 'lodash'
import os from 'os'
import path from 'path'
import process from 'process'
import dayjs from 'dayjs'
import fpmc from 'yf-fpm-client-js'
import axios from 'axios'

const LOCAL = path.join(__dirname, '..')
const CWD = process.cwd()
const DEV_MODE = LOCAL == CWD
let VIEW_ROOT_DIR = ''
if(!DEV_MODE){
    VIEW_ROOT_DIR = 'node_modules/fpm-plugin-admin/'
}

const SERVER_STATUS = {
    arch: os.arch(),
    cpus: os.cpus().length,
    hostname: os.hostname(),
    platform: os.platform(),
    type: os.type(),
    release: os.release(),
    freemem: Math.ceil(os.freemem() / 1024 / 1024 / 1024),
    totalmem: Math.ceil(os.totalmem() / 1024 / 1024 / 1024),
    uptime: os.uptime(),
}

export default (admin) => {
    admin.get('/admin', async (ctx) => {
        await ctx.render('admin/login.html')
    })
    
    admin.get('/admin/login', async (ctx) => {
        await ctx.render('admin/login.html')
    })
    
    admin.get('/admin/main', async (ctx) => {
        const startTime = dayjs(ctx.fpm._start_time).format('YYYY-MM-DD HH:mm:ss')
        await ctx.render('admin/main.html', {
            online: '24H',
            status: _.assign(SERVER_STATUS, {
                startTime,
                counter: ctx.fpm._counter,
            }),
        })
    })

    admin.get('/admin/test', async (ctx) => {
        await ctx.render('admin/test.html')
    })

    admin.get('/admin/config', async (ctx) => {
        await ctx.render('admin/config.html', {
            config: JSON.stringify(ctx.fpm.getConfig(), null, 2),
        })
    })

    admin.post('/admin/rpc', async (ctx) => {
        // get body
        const { method, args } = ctx.request.body
        try{
            const data = await new fpmc.Func(method).invoke(args)
            ctx.body = { errno: 0, data }
        }catch(e){
            ctx.body = e
        }
    })

    admin.post('/admin/registry', async (ctx) => {
        const { name } = ctx.request.body
        try{
            const data = await axios.get(`http://registry.npmjs.org/${name}`)
            const latest = data.data['dist-tags'].latest
            ctx.body = { errno: 0, data: { latest } }
        }catch(e){
            ctx.body = e
        }
    })

    admin.get('/admin/setting/:menu', async (ctx) => {
        await ctx.render('admin/setting/' + ctx.params.menu + '.html', {})
    })
    admin.get('/admin/about', async (ctx) => {
        await ctx.render('admin/about.html')
    })
    admin.get('/admin/plugin', async (ctx) => {
        // console.info(ctx.fpm.getPlugins())
        await ctx.render('admin/plugin.html', {
            plugins: ctx.fpm.getPlugins(),
        })
    })
    admin.get('/admin/logout', async (ctx) => {
        ctx.session.admin = undefined
        await ctx.redirect('/admin/login')
    })
    
    admin.post('/admin/login', async (ctx) => {
        //check pass
        let loginInfo = ctx.request.body
        const { user, pass, error} = _.assign({user: 'admin', pass: '741235896'}, ctx.fpm.getConfig('admin'))
        if (loginInfo.name === user && loginInfo.pass === pass) {
            // init fpmc
            const { port, hostname } = ctx.fpm.getConfig('server')
            fpmc.init({ appkey: '123123', masterKey: '123123', domain: `http://${ hostname || 'localhost'}:${port}`})
            // fpmc.ping().then(console.info).catch(console.error)
            ctx.session.admin = loginInfo
            ctx.body = { code: 0 }
        } else {
            ctx.body = { code: -99, error: error || '用户名或密码错误' }
        }
    })

    return admin
      
}